x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -236,11 +244,27 @@ private List<ShardSearchFailure> selectFailures() {
     private void sendOneNodeRequest(TargetShards targetShards, ComputeListener computeListener, NodeRequest request) {
         final ActionListener<List<DriverProfile>> listener = computeListener.acquireCompute();
         sendRequest(request.node, request.shardIds, request.aliasFilters, new NodeListener() {
+
+            private final Set<ShardId> pendingRetries = new HashSet<>();
+
             void onAfter(List<DriverProfile> profiles) {
                 nodePermits.get(request.node).release();
                 if (concurrentRequests != null) {
                     concurrentRequests.release();
                 }
+
+                if (pendingRetries.isEmpty() == false && remainingTargetShardSearchAttempts.getAndDecrement() > 0) {
+                    ongoingTargetShardResolutionAttempts.incrementAndGet();
+                    resolveShards(pendingRetries, computeListener.acquireAvoid().delegateFailure((l, resolutions) -> {
+                        for (var entry : resolutions.entrySet()) {
+                            targetShards.shards.get(entry.getKey()).remainingNodes.addAll(entry.getValue());
```

Comment:
Here we can execute the same target nodes up to 10 times if we hit failed on that node (with unavailable)?

Reply:
I do not follow. If the shard is unavailable it should no longer be in the routing table and should not be passed here.

Reply:
Here we are using the cluster state on the coordinator, which might not be up to dated.

Reply:
This is generally okay as we are going to retry more than once. 
We could make it transport action again and run on elected master by extending `TransportMasterNodeAction` but even then there is always a possibility of state changing while we waiting for the response.